vpr/src/pack/cluster.cpp

```diff
@@ -2522,6 +2562,50 @@ void add_cluster_molecule_candidates_by_highfanout_connectivity(t_pb* cur_pb,
     cur_pb->pb_stats->tie_break_high_fanout_net = AtomNetId::INVALID(); /* Mark off that this high fanout net has been considered */
 }
 
+void add_cluster_molecule_candidates_by_attraction_group(t_pb* cur_pb,
+                                                         t_cluster_placement_stats* cluster_placement_stats_ptr,
+                                                         const std::multimap<AtomBlockId, t_pack_molecule*>& atom_molecules,
+                                                         const int feasible_block_array_size) {
+    auto& atom_ctx = g_vpr_ctx.atom();
+    auto& floorplanning_ctx = g_vpr_ctx.mutable_floorplanning();
+    auto& attraction_info = floorplanning_ctx.attraction_groups;
+
+    //If the current cluster belongs to an attraction group, add all of the atoms
+    //from that attraction group to the feasible blocks
+    AttractGroupId grp_id = cur_pb->pb_stats->attraction_grp_id;
+    if (grp_id != NO_ATTRACTION_GROUP) {
+        AttractionGroup group = attraction_info.get_attraction_group_info(grp_id);
+
+        for (AtomBlockId blk_id : group.group_atoms) {
+            if (atom_ctx.lookup.atom_clb(blk_id) == ClusterBlockId::INVALID()) {
+                auto rng = atom_molecules.equal_range(blk_id);
+                for (const auto& kv : vtr::make_range(rng.first, rng.second)) {
```

Comment:
might be adding the molecule multiple times (because different atoms might be part of the same molecule) -> if not a problem, comment it, if it is -> can have an unordered set to check if it has been added.

Also put this in a subroutine and refactor all of the functions that go through this atom_molecules loop to call the subroutine

Reply:
Does not get added multiple times because add_molecule_to_pb_stats checks whether the molecule is already in feasible blocks